### PR TITLE
feat(core): integrate rspack_resolver with input_filesystem

### PR DESCRIPTION
**Pull request by hardfist**
related to #7478 
This PR support input_filesystem api in rust side and bridge with resolver's filesystem: this will not introduce performance regression

## Checklist

- [x] Tests updated (or not required).
- [x] Documentation updated (or not required).

### CODE REVIEW

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

**Comment by hardfist** [comment]
!bench

File: crates/rspack_fs/src/sync.rs [hunk 1]
```diff
@@ -23,11 +30,18 @@ pub trait WritableFileSystem {
...
-  /// Read the entire contents of a file into a bytes vector.
-  ///
-  /// Error: This function will return an error if path does not already exist.
-  fn read(&self, file: &Utf8Path) -> Result<Vec<u8>>;
+pub trait ReadableFileSystem: Send + Sync {
+  /// See [std::fs::read]
+  fn read(&self, path: &Path) -> io::Result<Vec<u8>>;
+
+  /// See [std::fs::metadata]
+  fn metadata(&self, path: &Path) -> io::Result<FileMetadata>;
```
Comment (h-a-n-a) [inline crates/rspack_fs/src/sync.rs:38]: I was a little bit confused. Why would this use structs defined in `rspack_resolver`, shouldn't this be a generic interface?

Comment (hardfist) [reply crates/rspack_fs/src/sync.rs:38]: you mean FileMetadata or FileSystem?

Comment (h-a-n-a) [reply crates/rspack_fs/src/sync.rs:38]: `FileMetadata`, sorry for not being clear.

Comment (hardfist) [reply crates/rspack_fs/src/sync.rs:38]: yes use std FileMetadata makes more sense, I will fix it

Comment (hardfist) [reply crates/rspack_fs/src/sync.rs]: fixed

**Comment by hardfist** [comment]
!bench

**Review by h-a-n-a — APPROVED** [review]

**Pull request merged**